### ISSUE TITLE
Wrong value passed to generic interface method in release

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/coreclr#18259]
On .NET core 2.1.0 and .NET framework 4.7.1 the following program prints `0` in debug but `1` in release:
```csharp
using System;

class C0
{
    public ulong F6;
    public C0(ulong f6)
    {
        F6 = f6;
    }
}

struct S0
{
    public C0 F2;
    public S0(C0 f2)
    {
        F2 = f2;
    }
}

struct S1
{
    public S0 F0;
    public S0 F2;
    public S1(S0 f2)
    {
        F0 = default(S0);
        F2 = f2;
    }
}

interface II0
{
    void WriteLine<T>(T b);
}

class C1 : II0
{
    public void WriteLine<T>(T b)
    {
        Console.WriteLine(b);
    }
}

public class Program
{
    private static II0 s_0;
    public static void Main()
    {
        s_0 = new C1();
        S1[] var0 = {new S1(new S0(new C0(0)))};
        var0[0].F2.F2.F6 = var0[0].F2.F2.F6;
        var0[0].F2.F2 = new C0(1); // The value specified here is printed in release
        var0[0].F2.F2.F6 = 0;
        s_0.WriteLine(var0[0].F2.F2.F6);
    }
}

```

**Comment by jakobbotsch** [comment]
I wasn't able to reduce this to a smaller example, unfortunately. However, the error does not occur in a few situations:
* If `s_0` is made a local
* If `var0` is not an array
* If `S1.F2` is made `C0` and a level of nesting removed in the bottom lines

**Comment by mikedn** [comment]
`var0[0].F2.F2.F6 = 0` tree post VN is
```
     ( 14, 14) [000111] ------------              *  STMT      void  (IL 0x06D...0x080)
N014 (  1,  1) [000108] ------------              |  /--*  CNS_INT   long   0 $c0
N015 ( 14, 14) [000110] -A-XG-------              \--*  ASG       long   $VN.Void
N013 ( 12, 12) [000109] ---XG--N----                 \--*  IND       long   $c0
N011 (  1,  1) [000294] ------------                    |  /--*  CNS_INT   long   8 field offset Fseq[c0_ul] $c1
N012 ( 10, 10) [000295] ----G--N----                    \--*  ADD       byref  <l:$383, c:$386>
N010 (  9,  9) [000106] *---G-------                       \--*  IND       ref    <l:$28e, c:$251>
N008 (  1,  1) [000296] ------------                          |  /--*  CNS_INT   long   8 field offset Fseq[s0_s1_1, s1_c0] $c1
N009 (  7,  7) [000297] ----G--N----                          \--*  ADD       byref  $483
N006 (  6,  6) [000308] ----G-------                             |  /--*  ADDR      byref  $480
N005 (  4,  4) [000102] a---G--N----                             |  |  \--*  IND       struct <l:$5c0, c:$504>
N003 (  1,  1) [000305] ------------                             |  |     |  /--*  CNS_INT   long   16 Fseq[#FirstElem] $c2
N004 (  2,  2) [000306] -------N----                             |  |     \--*  ADD       byref  $480
N002 (  1,  1) [000298] ------------                             |  |        \--*  LCL_VAR   ref    V03 tmp2         u:3 $340
N007 (  6,  6) [000307] ----G--N----                             \--*  COMMA     byref  $480
N001 (  0,  0) [000348] ------------                                \--*  NOP       void   $444
```
So the VN of the C0 reference where 0 is store is $28e. You'd expect this to be the VN produced by `new C0` above but it's not, it's the VN produce by the load in `var0[0].F2.F2.F6 = var0[0].F2.F2.F6;`. That is, VN missed the store `var0[0].F2.F2 = new C0(1);` and thinks that 0 is stored in the old C0 instance.

**Comment by mikedn** [comment]
The way the missing store is done is interesting:
```IL
	IL_005c: ldflda valuetype S0 S1::F2
	IL_0061: ldc.i4.1
	IL_0062: conv.i8
	IL_0063: newobj instance void C0::.ctor(uint64)
	IL_0068: stfld class C0 S0::F2
```
So the address of `var0[0].F2` is taken and then the new object is stored via `stdfld`.
```
     (  8,  8) [000094] ------------              *  STMT      void  (IL   ???...  ???)
N008 (  1,  1) [000279] ------------              |     /--*  CNS_INT   long   8 field offset Fseq[F2] $c1
N009 (  8,  8) [000280] ----G-------              |  /--*  ADD       byref  $482
N006 (  6,  6) [000291] ----G-------              |  |  |  /--*  ADDR      byref  $480
N005 (  4,  4) [000077] a---G--N----              |  |  |  |  \--*  IND       struct <l:$4c2, c:$503>
N003 (  1,  1) [000288] ------------              |  |  |  |     |  /--*  CNS_INT   long   16 Fseq[#FirstElem] $c2
N004 (  2,  2) [000289] -------N----              |  |  |  |     \--*  ADD       byref  $480
N002 (  1,  1) [000281] ------------              |  |  |  |        \--*  LCL_VAR   ref    V03 tmp2         u:3 $340
N007 (  6,  6) [000290] ----G--N----              |  |  \--*  COMMA     byref  $480
N001 (  0,  0) [000347] ------------              |  |     \--*  NOP       void   $443
N011 (  8,  8) [000093] -A--G---R---              \--*  ASG       byref  $482
N010 (  1,  1) [000092] D------N----                 \--*  LCL_VAR   byref  V08 tmp7         d:3 $482
…
     (  5,  4) [000099] ------------              *  STMT      void  (IL 0x068...  ???)
N003 (  1,  1) [000096] ------------              |  /--*  LCL_VAR   ref    V07 tmp6         u:3 (last use) $28f
N004 (  5,  4) [000098] -A-XG-------              \--*  ASG       ref    $VN.Void
N002 (  3,  2) [000097] *--XG--N----                 \--*  IND       ref    $28f
N001 (  1,  1) [000095] ------------                    \--*  LCL_VAR   byref  V08 tmp7         u:3 (last use) $482
```
I would guess that VN gets confused. The store address ($482) has field sequence [F2] and the store is supposed to be at `F2.F2`. But since the second F2 has offset 0 it disappeared into thin air.

That said, I'm not very familiar with this part of the JIT so it's not clear if VN is somehow wrong here or if the real problem is in how `stfld class C0 S0::F2` ends up being morphed in such a way that there's no associated field sequence.

**Comment by jakobbotsch** [comment]
If the problem is that the store disappears, isn't it strange that the signature of the interface method affects the bug? Do you know why it is sensitive to that?

**Comment by jakobbotsch** [comment]
I see, that makes sense. So the `CORINFO_HELP_VIRTUAL_FUNC_PTR` call inhibits the optimization that makes the result correct otherwise, even though the store is wrong in both cases?

**Comment by jakobbotsch** [comment]
This code produces the bug too:
```csharp
public static void Main()
{
    S1[] var0 = {new S1(new S0(new C0(0)))};
    var0[0].F2.F2.F6 = var0[0].F2.F2.F6;
    var0[0].F2.F2 = new C0(1); // The value specified here is printed in release
    var0[0].F2.F2.F6 = 0;
    GC.KeepAlive(var0);
    Console.WriteLine(var0[0].F2.F2.F6);
}
```
I assume because the JIT assumes that `GC.KeepAlive` could change `var0[0]`, so it has to reload it like before?

**Comment by mikedn** [comment]
> So the CORINFO_HELP_VIRTUAL_FUNC_PTR call inhibits the optimization that makes the result correct otherwise, even though the store is wrong in both cases?

Yes, but
* it probably should not, I don't think this particular helper call changes memory in an observable manner
* as you pretty much found out already, other calls can have the same effect. In general, any call that is not inlined is assumed to change memory in unspecified manner and that blocks CSEing of memory loads across call sites.

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

**Comment by jakobbotsch** [comment]
A bit simpler example:
```csharp
// Debug: Outputs 0
// Release: Outputs 2236837378
struct S1
{
    public uint F0;
    public S1(uint f0): this() { F0 = f0; }
}

struct S2
{
    public S1 F1;
    public S2(S1 f1): this() { F1 = f1; }
}

public class Program
{
    static S2[] s_11 = new S2[]{new S2(new S1(2236837378U))};
    public static void Main()
    {
        ref S1 vr7 = ref s_11[0].F1;
        vr7.F0 = vr7.F0;
        vr7.F0 = 0;
        vr7.F0 = vr7.F0;
        System.Console.WriteLine(vr7.F0);
    }
}
```

**Comment by briansull** [comment]
Investigating, This appears to be a different issue than #18672

**Comment by jakobbotsch** [comment]
@briansull should this be closed after #20838?

**Comment by briansull** [comment]
Yes, I will close this issue

Fixed with  #20838

**Comment by danmosemsft** [comment]
@briansull should this be reopened per above?

**Comment by briansull** [comment]
Should be a new issue as there can be multiple different issues in this area,

**Comment by jakobbotsch** [comment]
Done, opened #21231.
